### ISSUE TITLE
Checklist

### ISSUE THREAD

**Issue by shtefcs** [wekan/wekan#298]
Hi guys,

Awesome work I must say.

One of feature which I really miss and think other people would like to have it is _checklist_.

![](http://content.screencast.com/users/shtefcs/folders/Jing/media/17b8e2c1-a72a-4f7e-a210-c4f371e24135/2015-09-12_1837.png)

Do you plan to add it ?

Peace,

**Comment by mquandalle** [comment]
Yes, I plan to implement these for the next release.

**Comment by shtefcs** [comment]
Super :+1:

**Comment by zeineath** [comment]
:+1:

**Comment by bastianwegge** [comment]
:+1:

**Comment by aleks** [comment]
:+1:

**Comment by NickGodspeed** [comment]
+1

**Comment by skitoo** [comment]
:+1:

**Comment by skitoo** [comment]
:+1:

**Comment by Sohalt** [comment]
I'd like to be able to create checklists by using the GFM syntax (see https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments)

so when writing
```
[ ] foo
[x] bar
```

I'd like it to turn into a checklist of two items (foo and bar), with foo not checked and bar checked

using markdown for creating checklists also solves the question addressed in #129 because you can easily use markdowns `#` for creating multiple headings for checklists and also multiple checklists by simply spacing the `[ ]` items with newlines

https://github.com/FND/markdown-checklist might aid in implementing it.

**Comment by arctican** [comment]
I completely agree, checklists are probably the most important part of my workflow with Trello.

**Comment by shtefcs** [comment]
@mquandalle Maxime, can u give us checklist mate :) ? 

It past 1 year when u first start wondering do we need checklist. Yes, we really need.

I just imported trello board to wecan, but there is no checklist features, so it useles.

I would rather help you with that, but I just started learning Javascript.

In mean time, it would be really nice to provide us with checklist without thinking to much.

Peace!

**Comment by DaVo2k** [comment]
I would also vote for checklists! It seems to be the only feature preventing the use of wekan for me.

**Comment by shtefcs** [comment]
@DaVo2k Exactly!

**Comment by hsduk** [comment]
Much simpler than developing checklist ui components, pls see http://lifehacker.com/why-a-github-gist-is-my-favorite-to-do-list-1493063613
I just tried using gfm checkboxes as described in https://github.com/blog/1375%0A-task-lists-in-gfm-issues-pulls-comments in wekan, but did not work as the perak md package does not support full gfm. The summary feature is awesome imho.

- [ ] unchecked checkbox
- [X] checked checkbox

```
- [ ] for an empty checkbox
- [X] for an empty checkbox
```
Have raised https://github.com/perak/markdown/issues/7

**Comment by ousamabenyounes** [comment]
:+1: Would be happy to view this on the next release

**Comment by sunrez** [comment]
Another second for both checklists and nested tasks / sub-tasks, hierarchy .

**Comment by cgalibern** [comment]
:+1:

**Comment by bapthbrt** [comment]
:+1:

**Comment by ILM126** [comment]
:+1: 
I've got a friend that is holding off using Wekan until checklist is implemented :|

**Comment by shtefcs** [comment]
@ILM126 Same here :)

**Comment by Sohalt** [comment]
I have looked a little into this and sadly it seems that checklists can't easily be implemented without a change happening upstream in marked, the library that wekan uses for markdown translation (well, more precisely wekan uses perak:markdown, but that is just a meteor wrapper around marked). There is a pending [pull request](https://github.com/chjj/marked/pull/587) for checklist support, but it has not been merged yet.
Alternatively wekan could use a different library for markdown processing, such as [showdown](https://github.com/showdownjs/showdown) or [markdown-it](https://github.com/markdown-it/markdown-it), which provide checklists.
With the markdown part covered there still needs to be logic for saving and restoring the state, but I think that should be manageable.

**Comment by brylie** [comment]
It would be user friendly if checklists could be creates with point and click or touch, rather than text. Even though markdown seems simple, it is still quirky and less natural than pressing an "Add Checklist" button. In my opinion, people should not have to write markup of any kind when a UI element can provide enhanced usability.

**Comment by brylie** [comment]
@mquandalle how may I help with this feature? Is there a planning document or design patter that can be followed? Are there any sub-tasks towards creating a first-class checklist UI?

**Comment by Sohalt** [comment]
A GUI is in no way at odds with the markdown approach. I thought it would be easier to implement the markdown approach than adding additional GUI elements which is why I would prioritize it in order to quickly get checklists working at all, but I have nothing against implementing a GUI element *in addition* to the markdown.

**Comment by shtefcs** [comment]
Mark down approach would be best for MVP. It will give us functionality to have **check lists** until something more functional come.

**Comment by Sohalt** [comment]
I don't like the "until something more functional come" part. The markdown solution is very likely the fastest solution when having a dedicated keyboard since you are writing the card description anyways and a button means reaching for the mouse, positioning the cursor, clicking and repositioning the hands on the keyboard, which takes >2s. I am all for having a GUI for adding checkboxes so the casual user doesn't necessarily need to learn markdown, but this should never come at the expense of the power user since it is them who use the software extensively and use a feature hundreds of thousands times and every little bit of friction and every second lost add up. In this case especially I don't see any reason to remove the markdown solution once we've got a GUI for it, since the markdown in no way interferes with the GUI.

**Comment by shtefcs** [comment]
Then make it with GUI. For start, you can put similar icon below http://screencast.com/t/Sv2jmcmvJA

**Comment by brylie** [comment]
Here is one approach to how we could design the checklist component:

* checklists could be stored in a collection, 
* checklists would be associated with cards via 'foreign key': `checklistId` or `cardId`
* collection helper(s) would be added to the `Cards` or `Checklists` collection to retrieve related documents
* `checklistItems` would be an array of objects within each checklist document
  * each checklist item object could have a `completed` boolean attribute
  * It would be possible to add other attrubutes, such as `assignedTo`
* checklist progress could be calculated by array.length comparison between `completed: true` and `checklistItems.length`
* the component could be designed so that pressing "return" inserts a new checklist item and places the cursor into the text area, so no mouse clicks would be necessary

**Comment by brylie** [comment]
The collection based approach may save some effort in several areas:
 * we would not have to parse a markdown 'blob' for checklist markup
* we would not have to write any logic to map checklists to arrays. E.g. for comparing completed items with checklist length
* we would not have to wait on an upstream change in the markdown library

As mentioned before, the componentized UI approach may even be more efficient than typing `- [ ] ` for each checklist item.

**Comment by Sohalt** [comment]
Not very convinced about speed. Reaching for the mouse is a major obstacle. I assumed the "new item on enter" functionality as a given when going for a GUI approach (it's what trello does). Combined with a keyboard shortcut for starting a checklist it might be faster than markdown, but less flexible. For example you can't quickly insert an already checked item or mixed lists, e.g.
```
- foo
- [x] bar
- [ ] baz
```.
As for storage: the probably easiest solution would be to just store the markdown string, since you have to have that anyway when later editing the card. This might not be the most elegant solution but no extra model is required and the rendering is automatically being taken care of by the markdown library. The only issue is that the string needs to get updated when items get clicked

**Comment by Sohalt** [comment]
Also regarding not having to parse a "markdown blob" (whatever that exactly means): we still have to parse markdown since it is used as a way of structuring and styling card descriptions. When going for a dedicated storage model for checklists an additional complexity would be the need to store where in the description a checklist should be embedded (i.e. on which line).

**Comment by brylie** [comment]
One idea is that pressing the 'return' key would append a new checklist item, so that no mouse interaction would be required.

**Comment by brylie** [comment]
By 'markdown blob', I simply meant a string of markdown formatted text.

**Comment by brylie** [comment]
It also sounds like there are problems with the markdown library not supporting checklists, so rendering markdown-based checklists would be additional work.

**Comment by Sohalt** [comment]
Re: new item on enter: I got that (and am all for it). I just already assumed that would be the way to implement a dedicated checklist input since it appears to be the obvious way to do it (when opting for a GUI).

**Comment by brylie** [comment]
Cool, we sometimes need to state the obvious when planning to develop a feature.

**Comment by RobertasMurza** [comment]
@brylie [this topic](https://discuss.wekan.io/t/wekan-infinite-depth-architecture-board-child-board-card-checklist-and-subtask/55) on wekan forum also talks about checklist implementation.

**Comment by Sohalt** [comment]
The problems in the markdown library could be avoided by either switching to a different library (I've named to examples above) or temporarily forking the currently used one and patching checklist support (the PR already exists) and using that until it lands upstream.

**Comment by brylie** [comment]
@RobertasMurza thanks. I will read through that discussion.

In general, the component/collection approach could more naturally align with Meteor/JavaScript since the checklists would just be JSON documents. The markdown approach seems like it would take wrangling, such as regEx searches on the markdown string.

**Comment by Sohalt** [comment]
Not sure how advanced the markdown libraries are, but if the build an intermediary AST that could be extracted to avoid regex parsing or similar.

**Comment by brylie** [comment]
OK, cool. It still sounds similar though. In effect, we would be writing an intermediary layer to parse a data structure into a JavaScript object, when a collection/component approach would naturally store the data as JSON.

**Comment by brylie** [comment]
This sounds similar to the 'impedence mismatch' encountered in ORM libraries.

**Comment by Sohalt** [comment]
The proposal from the forum sounds tempting (I'm all for more simplicity and simply nesting cards in cards is about as simple/elegant as it gets), but I have major usability concerns. Allowing infinite recursion can quickly lead to an unmanageable mess with subtasks of subtasks of subtasks and so on. It also encourages overplanning/meta planning because you have to make decisions when to split something and what should be on what layer and for example at which granularity to assign due dates and people. The infinitely nested structure gets hard to visualize and it's easy to "get lost" somewhere in the hierarchy. The only reason I want checklists is because of the alternate UI. Checklist items are deliberately simple without any clutter, just a line of text with a checkbox in front, not a description, not a colored label, not a person assigned to it, not a due date on it. In my opinion checklists should not be for tracking long term progress, but clearly defined and quickly to carry out steps of a single conceptual activity (which is represented by the card). The obvious example would be a shopping list, where the card denotes the activity of going shopping and the list items are single things to purchase, which is  done in under a minute and doesn't need to be tracked on boards or assigned to people. Creating a new card for each item to purchase takes to long and clutters the board. A checklist on the other hand provides just the right amount of granularity in that it reminds you what to purchase and let's you track progress of that single coherent activity.

**Comment by whodafly** [comment]
This is the single thing that is preventing my open source project from moving away from Trello (which we really want to do). We can't move until checklists are implemented. I'm even happy to retype all the cards and checklists, I don't need the Trello importer. You guys have been talking about it for 6 months, and I've been watching and waiting, and I'll be honest, I'm a little frustrated. I've tried other open source solutions, and none of them matches what I want and need as much as Wekan does.

**Comment by shtefcs** [comment]
Maybe all we who needs checklist make some fond and collect money for it, then find someone who is willing to integrate ?

**Comment by rnjailamba** [comment]
There is something here https://github.com/wekan/wekan/pull/576
Why was it not merged

**Comment by wnereiz** [comment]
Waiting for this feature!

**Comment by brylie** [comment]
We are still waiting on feedback from @andersonlin in pull request #576 

It may be that @andersonlin does not have a Github account, so may not be able to sign the contributor agreement. Is there another way to move forward with that PR?

**Comment by shtefcs** [comment]
I sent email to Maxime to ask about Wekan future.

**Comment by brylie** [comment]
@mquandalle what can we do to get this feature,mand possibly PR #576, merged into Wekan?

**Comment by jelazos7** [comment]
I'm a little worried about the lack of GH activity from @mquandalle (none since Sep 2). If life has gotten in the way I can completely understand, but I wouldn't want to see Wekan die. (We cannot use Trello at my work place due to intellectual property and data protection policies.)

@brylie I would love to test out your checklist feature -- I've got my team hooked on Wekan and checklists is the missing thing to manage DoD and acceptance criteria. But, I had trouble getting your feature to work from source. Would you be willing to build a docker image and share?

**Comment by jelenajjo** [comment]
We would replace Trello completely if we had checklists.

**Comment by brylie** [comment]
@jelazos7 see pull request #630 for working checklist feature. Let us know what you think.

**Comment by Valiantiam** [comment]
What is the progress on this.

**Comment by ocdtrekkie** [comment]
@Valiantiam There is a pull request to add them, #630, but the maintainer appears to have abandoned the project. There is now a fork, at GitHub.com/wefork/wekan, and they have an issue to add this: https://github.com/wefork/wekan/issues/11 but as of yet nobody has moved forward on that.

**Comment by ocdtrekkie** [comment]
FYI, this feature has been merged.
